src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,180 @@
+
+-/
+
+/-- `affine_space` is an abbreviation for `add_torsor` in the
+case where the group is a vector space. -/
+@[nolint unused_arguments]
+abbreviation affine_space (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V]
+    [vector_space k V] :=
+add_torsor V P
+
+namespace affine_space
+
+open add_action
+open add_torsor
+
+variables (k : Type*) (V : Type*) {P : Type*} [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- The vector subspace spanning the differences of a (possibly empty)
+set of points. -/
+def vector_span (s : set P) : subspace k V := submodule.span k (vsub_set V s)
+
+/-- The points in the affine span of a (possibly empty) set of
+points. -/
+def span_points (s : set P) : set P :=
+{p | ∃ p1 ∈ s, ∃ v ∈ (vector_span k V s).carrier, p = v +ᵥ p1}
+
+/-- The set of points in the affine span of a nonempty set of points
+is nonempty. -/
+lemma span_points_nonempty_of_nonempty {s : set P} :
+  s.nonempty → (span_points k V s).nonempty 
+| ⟨p, hp⟩ := ⟨p, p, hp, 0, submodule.zero _, (zero_vadd V p).symm⟩
+
+/-- Adding a point in the affine span and a vector in the spanning
+subspace produces a point in the affine span. -/
+lemma vadd_mem_span_points_of_mem_span_points_of_mem_vector_span {s : set P} {p : P} {v : V}
+    (hp : p ∈ span_points k V s) (hv : v ∈ vector_span k V s) : v +ᵥ p ∈ span_points k V s :=
+begin
+  rcases hp with ⟨p2, ⟨hp2, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv2p, vadd_assoc],
+  use [p2, hp2, v + v2, (vector_span k V s).add hv hv2, rfl]
+end
+
+/-- Subtracting two points in the affine span produces a vector in the
+spanning subspace. -/
+lemma vsub_mem_vector_span_of_mem_span_points_of_mem_span_points {s : set P} {p1 p2 : P}
+    (hp1 : p1 ∈ span_points k V s) (hp2 : p2 ∈ span_points k V s) :
+  p1 -ᵥ p2 ∈ vector_span k V s :=
+begin
+  rcases hp1 with ⟨p1a, ⟨hp1a, ⟨v1, ⟨hv1, hv1p⟩⟩⟩⟩,
+  rcases hp2 with ⟨p2a, ⟨hp2a, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv1p, hv2p, vsub_vadd_eq_vsub_sub V (v1 +ᵥ p1a), vadd_vsub_assoc, add_comm, add_sub_assoc],
+  have hv1v2 : v1 - v2 ∈ (vector_span k V s).carrier,
+  { apply (vector_span k V s).add hv1,
+    rw ←neg_one_smul k v2,
+    exact (vector_span k V s).smul (-1 : k) hv2 },
+  refine (vector_span k V s).add _ hv1v2,
+  unfold vector_span,
+  change p1a -ᵥ p2a ∈ submodule.span k (vsub_set V s),
+  have hp1p2 : p1a -ᵥ p2a ∈ vsub_set V s, { use [p1a, hp1a, p2a, hp2a] },
+  have hp1p2s : vsub_set V s ⊆ submodule.span k (vsub_set V s) := submodule.subset_span,
+  apply set.mem_of_mem_of_subset hp1p2 hp1p2s
+end
+
+end affine_space
+
+open add_torsor affine_space
+
+section affine_subspace
+
+variables (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `vector_space k V`. -/
+structure affine_subspace :=
```

Comment:
Are you planning to define a lattice structure on `affine_subspace`? Probably `affine_span` is part of a galois connection, which might give you a bunch of "trivial lemmas" in a trivial way. Maybe a future PR?

Reply:
There's the complication that the intersection of two affine subspaces may be empty (but allowing affine subspaces to be empty would result in other complications, e.g. the choice of direction for such a subspace and defining its dimension). I suppose you could define e.g. a `semilattice_sup` on `affine_subspace`, and a lattice structure on `with_bot (affine_subspace k V P)`.

Reply:
Hmm, you are right. Let's ignore the lattice for the time being.